### PR TITLE
add option 'allowCirculatingSelection' to circulate selection

### PR DESCRIPTION
**Pull request by metasmile**
please confirm

### CODE REVIEW

File: DKImagePickerController/View/DKAssetGroupDetailVC.swift [hunk 1]
```diff
@@ -295,8 +295,13 @@ internal class DKAssetGroupDetailVC: UIViewController, UICollectionViewDelegate,
         }
         
         self.selectedGroupId = groupId
-		self.updateTitleView()
-		self.collectionView!.reloadData()
+        self.updateTitleView()
+        if(self.imagePickerController.deselectAllWhenChangingAlbum
+            || self.imagePickerController.allowCirculatingSelection){            
+            self.imagePickerController.deselectAllAssets()
+        }else{
```
Comment (zhangao0086) [inline DKImagePickerController/View/DKAssetGroupDetailVC.swift]:
```swift
if (...) {
} else {
}
```

File: DKImagePickerController/View/DKAssetGroupDetailVC.swift [hunk 2]
```diff
@@ -406,12 +411,26 @@ internal class DKAssetGroupDetailVC: UIViewController, UICollectionViewDelegate,
...
-		let shouldSelect = self.imagePickerController.selectedAssets.count < self.imagePickerController.maxSelectableCount
-		if !shouldSelect {
-			self.imagePickerController.UIDelegate.imagePickerControllerDidReachMaxLimit(self.imagePickerController)
-		}
-		
-		return shouldSelect
+        let didReachMaxSelectableCount = self.imagePickerController.selectedAssets.count >= self.imagePickerController.maxSelectableCount
+        if didReachMaxSelectableCount {
+            if self.imagePickerController.allowCirculatingSelection {
+                let currentSelectedIndexPaths:[NSIndexPath] = collectionView.indexPathsForSelectedItems()!
```
Comment (zhangao0086) [inline DKImagePickerController/View/DKAssetGroupDetailVC.swift]: `let currentSelectedIndexPaths : [NSIndexPath] = collectionView.indexPathsForSelectedItems()!`

File: DKImagePickerController/View/DKAssetGroupDetailVC.swift [hunk 3]
```diff
@@ -406,12 +411,26 @@ internal class DKAssetGroupDetailVC: UIViewController, UICollectionViewDelegate,
...
+                let firstSelectedIndexPath:NSIndexPath = currentSelectedIndexPaths.first!
+                if let removingAsset:DKAsset = (collectionView.cellForItemAtIndexPath(firstSelectedIndexPath) as! DKAssetCell).asset{
+                    self.imagePickerController.deselectImage(removingAsset)
+                }
+                UIView.performWithoutAnimation({
+                    collectionView.reloadItemsAtIndexPaths(currentSelectedIndexPaths)
+                })
+                return true
+                
+            }else{
```
Comment (zhangao0086) [inline DKImagePickerController/View/DKAssetGroupDetailVC.swift]: `} else {`

File: DKImagePickerController/View/DKAssetGroupDetailVC.swift [hunk 4]
```diff
@@ -406,12 +411,26 @@ internal class DKAssetGroupDetailVC: UIViewController, UICollectionViewDelegate,
...
-		if !shouldSelect {
-			self.imagePickerController.UIDelegate.imagePickerControllerDidReachMaxLimit(self.imagePickerController)
-		}
-		
-		return shouldSelect
+        let didReachMaxSelectableCount = self.imagePickerController.selectedAssets.count >= self.imagePickerController.maxSelectableCount
+        if didReachMaxSelectableCount {
+            if self.imagePickerController.allowCirculatingSelection {
+                let currentSelectedIndexPaths:[NSIndexPath] = collectionView.indexPathsForSelectedItems()!
+                let firstSelectedIndexPath:NSIndexPath = currentSelectedIndexPaths.first!
```
Comment (zhangao0086) [inline DKImagePickerController/View/DKAssetGroupDetailVC.swift]: `let firstSelectedIndexPath : NSIndexPath`

File: DKImagePickerController/View/DKAssetGroupDetailVC.swift [hunk 5]
```diff
@@ -406,12 +411,26 @@ internal class DKAssetGroupDetailVC: UIViewController, UICollectionViewDelegate,
...
+                if let removingAsset:DKAsset = (collectionView.cellForItemAtIndexPath(firstSelectedIndexPath) as! DKAssetCell).asset{
```
Comment (zhangao0086) [inline DKImagePickerController/View/DKAssetGroupDetailVC.swift]: `if let removingAsset : DKAsset = ...`

**Comment by metasmile** [comment]
Sorry I did cancel this pull request. forked repo will be separated.

**Pull request closed without merging**